Add NVFP4 two-level scaling to scaled_mm

slayton58 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #165774

Summary:

* Add second-level scaling dispatch to scaled_mm, tying into optional `alpha` passing
* Add two-level tests

Test Plan:

```
pytest -svv -k "nvfp4_global_scale" test/test_scaled_matmul_cuda.py
```

Reviewers:

Subscribers:

Tasks:

Tags:
Signed-off-by: Simon Layton <simonlayton@meta.com>

slayton58 committed:
Add NVFP4 two-level scaling to scaled_mm

Summary:

* Add second-level scaling dispatch to scaled_mm, tying into optional `alpha` passing
* Add two-level tests

Test Plan:

```
pytest -svv -k "nvfp4_global_scale" test/test_scaled_matmul_cuda.py
```

Reviewers:

Subscribers:

Tasks:

Tags:
Signed-off-by: Simon Layton <simonlayton@meta.com>

[ghstack-poisoned]

slayton58 committed:
Update on "Add NVFP4 two-level scaling to scaled_mm"


Summary:

* Add second-level scaling dispatch to scaled_mm, tying into optional `alpha` passing
* Add two-level tests

Test Plan:

```
pytest -svv -k "nvfp4_global_scale" test/test_scaled_matmul_cuda.py
```

Reviewers:

Subscribers:

Tasks:

Tags:
Signed-off-by: Simon Layton <simonlaytonmeta.com>

[ghstack-poisoned]

slayton58 commented on test/test_scaled_matmul_cuda.py:
```diff
@@ -1254,6 +1290,59 @@ def test_pack_uint4(self):
+        (2, 1024, 128),
+        (127, 96, 1024),
+        (1025, 128, 96)
+    ], name_fn=lambda mkn: f"{mkn[0]}_{mkn[1]}_{mkn[2]}")
```
Nit: this is directly from the `test_blockwise_mxfp8_nvfp4_mxfp4_numerics` test parametrize annotations - I can change both.

slayton58 committed:
Update on "Add NVFP4 two-level scaling to scaled_mm"


Summary:

* Add second-level scaling dispatch to scaled_mm, tying into optional `alpha` passing
* Add two-level tests

Test Plan:

```
pytest -svv -k "nvfp4_global_scale" test/test_scaled_matmul_cuda.py
```

Reviewers:

Subscribers:

Tasks:

Tags:
Signed-off-by: Simon Layton <simonlaytonmeta.com>

[ghstack-poisoned]

drisspg commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -2322,12 +2322,20 @@ _scaled_nvfp4_nvfp4(
 #endif
-  TORCH_CHECK_VALUE(single_scale, "Only single-scaled NVFP4 currently supported");
+  std::optional<Tensor> alpha = std::nullopt;
+  if (global_scale_a.has_value() && global_scale_b.has_value()) {
```
should this be scale_a or scale_b?

slayton58 replied:
I see - if one global scale is passed, trigger the check for both scales present, as opposed to now, where if you pass a single global scale, it'll silently do nothing with it.

slayton58 committed:
Update on "Add NVFP4 two-level scaling to scaled_mm"


Summary:

* Add second-level scaling dispatch to scaled_mm, tying into optional `alpha` passing
* Add two-level tests

Test Plan:

```
pytest -svv -k "nvfp4_global_scale" test/test_scaled_matmul_cuda.py
```

Reviewers:

Subscribers:

Tasks:

Tags:
Signed-off-by: Simon Layton <simonlaytonmeta.com>

[ghstack-poisoned]

slayton58 commented:
@pytorchbot rebase -b main

slayton58 committed:
Update

[ghstack-poisoned]
